api/src/middleware/error-handler.ts

```diff
@@ -1,105 +1,125 @@
-import { ErrorCode, MethodNotAllowedError, isDirectusError } from '@directus/errors';
-import { isObject, toArray } from '@directus/utils';
+import { ErrorCode, isDirectusError } from '@directus/errors';
+import { isObject } from '@directus/utils';
 import { getNodeEnv } from '@directus/utils/node';
-import type { ErrorRequestHandler } from 'express';
+import type { ErrorRequestHandler, NextFunction, Request, Response } from 'express';
 import getDatabase from '../database/index.js';
 import emitter from '../emitter.js';
 import { useLogger } from '../logger.js';
 
-// Note: keep all 4 parameters here. That's how Express recognizes it's the error handler, even if
-// we don't use next
-const errorHandler: ErrorRequestHandler = (err, req, res, _next) => {
-	const logger = useLogger();
-
-	let payload: any = {
-		errors: [],
+type ApiError = {
+	message: string;
+	extensions: {
+		code: string;
+		[key: string]: any;
 	};
+};
 
-	const errors = toArray<unknown>(err);
+const FALLBACK_ERROR = {
+	status: 500,
+	message: 'An unexpected error occurred.',
+	extensions: {
+		code: 'INTERNAL_SERVER_ERROR',
+	},
+};
 
+export const errorHandler = asyncErrorHandler(async (err, req, res) => {
+	const logger = useLogger();
+
+	let errors: ApiError[] = [];
 	let status: number | null = null;
 
-	for (const error of errors) {
+	// It can be assumed that at least one error is given
+	const receivedErrors: unknown[] = Array.isArray(err) ? err : [err];
+
+	for (const error of receivedErrors) {
 		if (getNodeEnv() === 'development') {
-			if (isObject(error)) {
-				error['extensions'] = {
-					...(error['extensions'] || {}),
-					stack: error['stack'],
-				};
+			// If available, expose stack trace under error's extensions data
+			if (isObject(error) && error['stack']) {
+				(error['extensions'] ??= {} as any).stack = error['stack'];
```

Comment:
Normally I'm a fan of short handed assignments, but maybe this is more readable?

```suggestion
			if (isObject(error) && error['stack'] && (error['extensions'] === undefined || isObject(error['extensions']))) {
				error['extensions'] = {
					...error['extensions'],
					stack: error['stack'],
				}
```